From 6b7d8811757f3d2c99ffdc28e67dd6a4ffcc3d1b Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Tue, 9 Apr 2019 18:14:44 +0100
Subject: [PATCH 394/587] drm: vc4: Need to call drm_crtc_vblank_[on|off] from
 vc4_crtc_[en|dis]able

vblank needs to be enabled and disabled by the driver to avoid the
DRM framework complaining in the kernel log.

vc4_fkms_disable_vblank needs to signal that we don't want vblank
callbacks too.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
---
 drivers/gpu/drm/vc4/vc4_firmware_kms.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/gpu/drm/vc4/vc4_firmware_kms.c b/drivers/gpu/drm/vc4/vc4_firmware_kms.c
index a82166943555..71fbf46e0e8a 100644
--- a/drivers/gpu/drm/vc4/vc4_firmware_kms.c
+++ b/drivers/gpu/drm/vc4/vc4_firmware_kms.c
@@ -562,6 +562,8 @@ static void vc4_crtc_mode_set_nofb(struct drm_crtc *crtc)
 
 static void vc4_crtc_disable(struct drm_crtc *crtc, struct drm_crtc_state *old_state)
 {
+	drm_crtc_vblank_off(crtc);
+
 	/* Always turn the planes off on CRTC disable. In DRM, planes
 	 * are enabled/disabled through the update/disable hooks
 	 * above, and the CRTC enable/disable independently controls
@@ -577,6 +579,7 @@ static void vc4_crtc_disable(struct drm_crtc *crtc, struct drm_crtc_state *old_s
 
 static void vc4_crtc_enable(struct drm_crtc *crtc, struct drm_crtc_state *old_state)
 {
+	drm_crtc_vblank_on(crtc);
 	/* Unblank the planes (if they're supposed to be displayed). */
 
 	if (crtc->primary->state->fb)
@@ -673,6 +676,9 @@ static int vc4_fkms_enable_vblank(struct drm_crtc *crtc)
 
 static void vc4_fkms_disable_vblank(struct drm_crtc *crtc)
 {
+	struct vc4_crtc *vc4_crtc = to_vc4_crtc(crtc);
+
+	vc4_crtc->vblank_enabled = false;
 }
 
 static const struct drm_crtc_funcs vc4_crtc_funcs = {
-- 
2.17.1

